ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -0,0 +1,822 @@
+          , CC.Block.cvsUtxo         = utxo
+          , CC.Block.cvsUpdateState  = updateState
+          , CC.Block.cvsDelegationState = delegationState
+          }
+        snapshots' = trimSnapshots $
+          if (CC.Block.cvsDelegationState state' == CC.Block.cvsDelegationState state)
+          then snapshots
+          else
+            let startOfSnapshot = case snapshots of
+                  _ Seq.:|> a -> sbUpper a
+                  Seq.Empty   -> SlotNo 0
+            in snapshots Seq.|> slotBounded startOfSnapshot (convertSlot $ CC.Block.blockSlot block) state'
+
+      pure $ ByronLedgerState state' snapshots'
+    where
+      bodyState = CC.Block.BodyState
+        { CC.Block.utxo        = CC.Block.cvsUtxo state
+        , CC.Block.updateState = CC.Block.cvsUpdateState state
+        , CC.Block.delegationState = CC.Block.cvsDelegationState state
+        }
+      bodyEnv = CC.Block.BodyEnvironment
+        { CC.Block.protocolMagic = fixPM $ Genesis.configProtocolMagic cfg
+        , CC.Block.k = Genesis.configK cfg
+        , CC.Block.numGenKeys
+        , CC.Block.protocolParameters = CC.UPI.adoptedProtocolParameters . CC.Block.cvsUpdateState $ state
+        , CC.Block.currentEpoch = CC.Slot.slotNumberEpoch (Genesis.configEpochSlots cfg) (CC.Block.blockSlot block)
+        }
+      numGenKeys :: Word8
+      numGenKeys =
+        case length (Genesis.unGenesisWStakeholders $ Genesis.configBootStakeholders cfg) of
+          n
+            | n > fromIntegral (maxBound :: Word8) -> panic
+              "updateBody: Too many genesis keys"
+            | otherwise -> fromIntegral n
+      fixPM (Crypto.AProtocolMagic a b) = Crypto.AProtocolMagic (reAnnotate a) b
+      trimSnapshots = Seq.dropWhileL (\ss -> sbUpper ss
+                                       < convertSlot (CC.Block.blockSlot block) - 2*(coerce $ Genesis.configK cfg))
+
+  applyLedgerHeader (ByronLedgerConfig cfg) (ByronBlock block) (ByronLedgerState state snapshots)
+    = mapExcept (bimap ByronLedgerError (\i -> ByronLedgerState i snapshots)) $ do
+      updateState <- CC.Block.updateHeader headerEnv (CC.Block.cvsUpdateState state) (CC.Block.blockHeader block)
+      pure $ state
+        { CC.Block.cvsLastSlot     = CC.Block.blockSlot block
+        , CC.Block.cvsPreviousHash = Right $ CC.Block.blockHashAnnotated block
+        , CC.Block.cvsUpdateState  = updateState
+        }
+    where
+      headerEnv = CC.Block.HeaderEnvironment
+        { CC.Block.protocolMagic = fixPMI $ Genesis.configProtocolMagicId cfg
+        , CC.Block.k          = Genesis.configK cfg
+        , CC.Block.numGenKeys
+        , CC.Block.delegationMap
+        , CC.Block.lastSlot   = CC.Block.cvsLastSlot state
+        }
+      numGenKeys :: Word8
+      numGenKeys =
+        case length (Genesis.unGenesisWStakeholders $ Genesis.configBootStakeholders cfg) of
+          n
+            | n > fromIntegral (maxBound :: Word8) -> panic
```

Comment:
Let's use `error`; if we decide that `panic` is the better option, we should do so throughout, not just in one place.